--- setup.py.orig	2010-07-11 19:25:09.323393818 -0500
+++ setup.py	2010-07-11 19:25:33.380050367 -0500
@@ -6,18 +6,18 @@
 import string
 import re
 import __builtin__
-import distutils.core as core
-
-from os.path import join as pathjoin
-from distutils.command.install import install as _install
-from distutils.command.sdist import sdist as _sdist
-from distutils.command.build_scripts import build_scripts as _build_scripts
 try:
     import test.regrtest as regrtest
 except ImportError, e:
     print >> sys.stderr, ('Warning: setup.py cannot be used for running the'
                                     ' regression test suite on this platform')
     print >> sys.stderr, 'ImportError: %s' % e
+import distutils.core as core
+
+from os.path import join as pathjoin
+from distutils.command.install import install as _install
+from distutils.command.sdist import sdist as _sdist
+from distutils.command.build_scripts import build_scripts as _build_scripts
 
 import pyclewn_install
 from clewn import *
