--- webmin-1.410.orig/ldap-useradmin/save_user.cgi	2008-04-25 12:22:46.000000000 +0200
+++ webmin-1.410/ldap-useradmin/save_user.cgi	2008-04-25 12:23:25.000000000 +0200
@@ -345,7 +345,7 @@ else {
 		$shadow = &shadow_fields();
 
 		# Add to the ldap database
-		@classes = ( "posixAccount", "shadowAccount" );
+		@classes = ( "posixAccount", "shadowAccount", "person" );
 		if ($schema->objectclass("person") && $config{'person'}) {
 			push(@classes, "person");
 			}
