diff -dNur src.orig/arch/MovieTexture/MovieTexture_FFMpeg.cpp src/arch/MovieTexture/MovieTexture_FFMpeg.cpp
--- src.orig/arch/MovieTexture/MovieTexture_FFMpeg.cpp	2009-05-09 11:17:08.000000000 -0400
+++ src/arch/MovieTexture/MovieTexture_FFMpeg.cpp	2009-05-09 11:18:43.000000000 -0400
@@ -35,15 +35,15 @@
 #include <libswscale/swscale.h>
 }
 #endif // HAVE_IMG_CONVERT
-    void img_convert__(AVPicture *dst, int dst_pix_fmt,
-                       const AVPicture *src, int pix_fmt,
+    void img_convert__(AVPicture *dst, enum PixelFormat dst_pix_fmt,
+                       const AVPicture *src, enum PixelFormat pix_fmt,
                        int width, int height)
     {
 #ifdef HAVE_IMG_CONVERT
         img_convert(dst, dst_pix_fmt, src, pix_fmt, width, height);
 #else
         static SwsContext* context = 0;
-        context = sws_getCachedContext(context, width, height, pix_fmt, width, height, dst_pix_fmt, 0, NULL, NULL, NULL);
+        context = sws_getCachedContext(context, width, height, pix_fmt, width, height, dst_pix_fmt, SWS_SPLINE, NULL, NULL, NULL);
         sws_scale(context, const_cast<uint8_t**>(src->data), const_cast<int*>(src->linesize), 0, height, dst->data, dst->linesize);
 #endif // HAVE_IMG_CONVERT
     }
