--- CMakeLists.txt.orig	2008-04-17 21:46:43.000000000 -0500
+++ CMakeLists.txt	2008-04-17 21:47:22.000000000 -0500
@@ -385,7 +385,7 @@
 
 OPTION(ITK_USE_LIBXML2 "Use LIBXML2." OFF)
 MARK_AS_ADVANCED(ITK_USE_LIBXML2)
-ITK_THIRD_PARTY_OPTION(LIBXML2 libxml2)
+ITK_THIRD_PARTY_OPTION(LibXml2 libxml2)
 
 #-----------------------------------------------------------------------------
 # Dispatch the build into the proper subdirectories.
