--- ./gvpe-2.21/doc/complex-example/gvpe.conf_old	2008-09-13 17:11:55.727708113 -0400
+++ ./gvpe-2.21/doc/complex-example/gvpe.conf	2008-09-13 17:14:21.637708102 -0400
@@ -32,107 +32,107 @@
 # node-specific.
 
 # marc@lap
-node = mobil
+#node = mobil
 
 # marc@home
-node = doom
-enable-rawip = yes
-enable-tcp = yes
+#node = doom
+#enable-rawip = yes
+#enable-tcp = yes
 
 # marc@uni
-node = ruth
-enable-rawip = yes
-enable-tcp = yes
-enable-icmp = yes
-hostname = 200.100.162.95
-connect = always
-router-priority = 30
-on ruth node-up = node-up
-on ruth node-down = node-up
+#node = ruth
+#enable-rawip = yes
+#enable-tcp = yes
+#enable-icmp = yes
+#hostname = 200.100.162.95
+#connect = always
+#router-priority = 30
+#on ruth node-up = node-up
+#on ruth node-down = node-up
 
 # marc@mu
-node = frank
-enable-rawip = yes
-hostname = 44.88.167.250
-router-priority = 20
-connect = always
+#node = frank
+#enable-rawip = yes
+#hostname = 44.88.167.250
+#router-priority = 20
+#connect = always
 
 # nethype
-node = rain
-enable-rawip = yes
-hostname = 145.253.105.130
-router-priority = 10
-connect = always
+#node = rain
+#enable-rawip = yes
+#hostname = 145.253.105.130
+#router-priority = 10
+#connect = always
 
 # marco@home
-node = marco
-enable-rawip = yes
+#node = marco
+#enable-rawip = yes
 
 # stefan@ka
-node = wappla
-connect = never
+#node = wappla
+#connect = never
 
 # stefan@lap
-node = stefan
-udp-port = 408
-connect = never
+#node = stefan
+#udp-port = 408
+#connect = never
 
 # paul@wg
-node = n8geil
-on ruth enable-icmp = yes
-on n8geil enable-icmp = yes
-enable-udp = no
+#node = n8geil
+#on ruth enable-icmp = yes
+#on n8geil enable-icmp = yes
+#enable-udp = no
 
 # paul@lap
-node = syrr
+#node = syrr
 
 # paul@lu
-node = donomos
+#node = donomos
 
 # marco@hn
-node = core
+#node = core
 
 # elmex@home
-node = elmex
-enable-rawip = yes
-hostname = 100.251.143.181
+#node = elmex
+#enable-rawip = yes
+#hostname = 100.251.143.181
 
 # stefan@kwc.at
-node = fwkw
-connect = never
-on stefan connect = always
-on wappla connect = always
-hostname = 182.73.81.146
+#node = fwkw
+#connect = never
+#on stefan connect = always
+#on wappla connect = always
+#hostname = 182.73.81.146
 
 # elmex@home
-node = jungfrau
-enable-rawip = yes
+#node = jungfrau
+#enable-rawip = yes
 
 # uni main router
-node = surfer
-enable-rawip = yes
-enable-tcp = no
-enable-icmp = yes
-hostname = 200.100.162.79
-connect = always
-router-priority = 40
+#node = surfer
+#enable-rawip = yes
+#enable-tcp = no
+#enable-icmp = yes
+#hostname = 200.100.162.79
+#connect = always
+#router-priority = 40
 
 # jkneer@marvin
-node = marvin
-enable-rawip = yes
-enable-udp = no
+#node = marvin
+#enable-rawip = yes
+#enable-udp = no
 
 # jkneer@entrophy
-node = entrophy
-enable-udp = no
-enable-tcp = yes
+#node = entrophy
+#enable-udp = no
+#enable-tcp = yes
 
 # mr. primitive
-node = voyager
-enable-udp = no
-enable-tcp = no
-on voyager enable-tcp = yes
-on voyager enable-udp = yes
+#node = voyager
+#enable-udp = no
+#enable-tcp = no
+#on voyager enable-tcp = yes
+#on voyager enable-udp = yes
 
 # v-server (barbados.dn-systems.de)
 #node = vserver
