--- ../utf8-to-roff	2008-03-10 21:20:51.000000000 -0700
+++ ./utf8-to-roff	2008-12-17 17:33:57.000000000 -0800
@@ -1,6 +1,8 @@
-#! /usr/bin/perl -w -CI
+if (true || '\'); then exec perl -CI "$0" "$@"; fi  # ') {}
+
+# #! /usr/bin/perl -w -CI
 # Converts a man page from utf8 (not understood by groff 1.18) to groff escapes.
 # I couldn't find any existing tool to convert from utf8 to groff, though I
 # seem to recall seeing some code to do so somewhere.
 #
 # Hereby released into public domain by Peter Moulder.
