--- ../Makefile.am	2008-12-17 20:32:33.000000000 -0800
+++ src/Makefile.am	2008-12-17 20:34:53.000000000 -0800
@@ -23,12 +23,12 @@
 	$(XFT_CFLAGS)	\
 	$(POPPLER_CFLAGS)	\
 	$(POPPLER_GLIB_CFLAGS)	\
 	-DPOTRACE=\"potrace\"	\
 	$(INKSCAPE_CFLAGS) \
-	-I$(top_srcdir)/cxxtest
-
+	-I$(top_srcdir)/cxxtest \
+	-I$(top_srcdir)/src/extension/script
 include Makefile_insert
 include application/Makefile_insert
 include dialogs/Makefile_insert
 include display/Makefile_insert
 include dom/Makefile_insert
