diff -ur mysql-6.0.9-alpha/Makefile.am mysql-6.0.9-alpha_new/Makefile.am
--- mysql-6.0.9-alpha/Makefile.am	2009-01-02 10:45:30.000000000 -0500
+++ mysql-6.0.9-alpha_new/Makefile.am	2009-01-30 22:39:48.000000000 -0500
@@ -32,7 +32,7 @@
 			@sql_union_dirs@ unittest storage plugin \
 			@sql_server@ @man_dirs@ tests \
 			netware @libmysqld_dirs@ \
-			mysql-test support-files sql-bench \
+			support-files \
 			win
 
 DIST_SUBDIRS =		. include Docs zlib \
@@ -41,7 +41,7 @@
 			strings mysys dbug mysys/tests extra regex libmysql libmysql_r client unittest storage plugin \
 			vio sql man tests \
 			netware libmysqld \
-			mysql-test support-files sql-bench \
+			support-files \
 			win \
 			BUILD
 DISTCLEANFILES = ac_available_languages_fragment
@@ -93,41 +93,6 @@
 test-unit-big:
 	cd unittest; MYTAP_CONFIG=big $(MAKE) test
 
-test-ps:
-	cd mysql-test ; \
-	    @PERL@ ./mysql-test-run.pl $(force) --ps-protocol --mysqld=--binlog-format=mixed
-
-test-nr:
-	cd mysql-test ; \
-	    @PERL@ ./mysql-test-run.pl $(force) --mysqld=--binlog-format=row
-
-test-pr:
-	cd mysql-test ; \
-	    @PERL@ ./mysql-test-run.pl $(force) $(mem) --ps-protocol --mysqld=--binlog-format=row #@libevent_test_option@
-
-test-ns:
-	cd mysql-test ; \
-	    @PERL@ ./mysql-test-run.pl $(force) $(mem) --mysqld=--binlog-format=mixed
-
-test-binlog-statement:
-	cd mysql-test ; \
-	    @PERL@ ./mysql-test-run.pl $(force) --mysqld=--binlog-format=statement
-
-# This code is duplicated in "test-bt", see the Changeset comment of 2007-Dec-07
-test-embedded:
-	if [ -e bin/mysqltest_embedded -o -e libmysqld/examples/mysqltest_embedded ] ; then \
-	  cd mysql-test ; MTR_BUILD_THREAD=auto \
-	      @PERL@ ./mysql-test-run.pl --comment=embedded --force --timer \
-	          --embedded-server --skip-rpl --skip-ndbcluster ; \
-	else \
-	  echo "no program found for 'embedded' tests - skipped testing" ; \
-	fi
-
-test-reprepare:
-	cd mysql-test ; \
-	    @PERL@ ./mysql-test-run.pl $(force) $(mem) --ps-protocol \
-		--mysqld=--debug=+d,reprepare_each_statement
-
 test:	test-ns test-pr
 
 test-full:	test test-nr test-ps
@@ -142,73 +107,6 @@
 test-force-mem:
 	$(MAKE) force=--force mem=--mem test
 
-test-bt:
-	-cd mysql-test ; MTR_BUILD_THREAD=auto \
-	    @PERL@ ./mysql-test-run.pl --comment=normal --force --timer \
-	        --skip-ndbcluster --report-features
-	-cd mysql-test ; MTR_BUILD_THREAD=auto \
-	    @PERL@ ./mysql-test-run.pl  --comment=ps    --force --timer \
-	        --skip-ndbcluster --ps-protocol
-	-if [ -e bin/ndbd -o -e storage/ndb/src/kernel/ndbd ] ; then \
-	  cd mysql-test ; \
-	    MTR_BUILD_THREAD=auto \
-	      @PERL@ ./mysql-test-run.pl --comment=ndb+rpl_ndb+ps --force --timer \
-	          --ps-protocol --mysqld=--binlog-format=row --suite=ndb,rpl_ndb ; \
-	    MTR_BUILD_THREAD=auto \
-	      @PERL@ ./mysql-test-run.pl --comment=ndb --force --timer \
-	          --with-ndbcluster-only ; \
-	else \
-	  echo "no program found for 'ndbcluster' tests - skipped testing" ; \
-	fi
-	-cd mysql-test ; MTR_BUILD_THREAD=auto \
-	    @PERL@ ./mysql-test-run.pl --force --comment=falcon --suite=falcon
-	-cd mysql-test ; MTR_BUILD_THREAD=auto \
-	    @PERL@ ./mysql-test-run.pl --force --comment=funcs1+ps --ps-protocol --reorder --suite=funcs_1
-	-cd mysql-test ; MTR_BUILD_THREAD=auto \
-	    @PERL@ ./mysql-test-run.pl --force --comment=funcs2 --suite=funcs_2
-	-cd mysql-test ; MTR_BUILD_THREAD=auto \
-	    @PERL@ ./mysql-test-run.pl --force --comment=partitions --suite=parts
-	-cd mysql-test ; MTR_BUILD_THREAD=auto \
-	    @PERL@ ./mysql-test-run.pl --force --comment=stress --suite=stress
-	-cd mysql-test ; MTR_BUILD_THREAD=auto \
-	    @PERL@ ./mysql-test-run.pl --force --comment=jp --suite=jp
-	-if [ -d mysql-test/suite/nist ] ; then \
-	  cd mysql-test ; MTR_BUILD_THREAD=auto \
-	      @PERL@ ./mysql-test-run.pl --comment=nist --force --suite=nist ; \
-	fi
-	-if [ -d mysql-test/suite/nist ] ; then \
-	  cd mysql-test ; MTR_BUILD_THREAD=auto \
-	      @PERL@ ./mysql-test-run.pl --comment=NIST+normal --force --suite=nist ; \
-	      @PERL@ ./mysql-test-run.pl --comment=NIST+ps --force --suite=nist --ps-protocol ; \
-	fi
-	-if [ -e bin/mysqltest_embedded -o -e libmysqld/examples/mysqltest_embedded ] ; then \
-	  cd mysql-test ; MTR_BUILD_THREAD=auto \
-	      @PERL@ ./mysql-test-run.pl --comment=embedded --force --timer \
-	          --embedded-server --skip-rpl --skip-ndbcluster ; \
-	else \
-	  echo "no program found for 'embedded' tests - skipped testing" ; \
-	fi
-
-test-bt-fast:
-	-cd mysql-test ; MTR_BUILD_THREAD=auto \
-	    @PERL@ ./mysql-test-run.pl  --comment=ps    --force --timer \
-	        --skip-ndbcluster --ps-protocol --report-features
-	-if [ -e bin/ndbd -o -e storage/ndb/src/kernel/ndbd ] ; then \
-	  cd mysql-test ; \
-	    MTR_BUILD_THREAD=auto \
-	      @PERL@ ./mysql-test-run.pl --comment=ndb --force --timer \
-	          --with-ndbcluster-only ; \
-	else \
-	  echo "no program found for 'ndbcluster' tests - skipped testing" ; \
-	fi
-	-cd mysql-test ; MTR_BUILD_THREAD=auto \
-	    @PERL@ ./mysql-test-run.pl --force --comment=stress --suite=stress
-
-test-bt-debug:
-	-cd mysql-test ; MTR_BUILD_THREAD=auto \
-	    @PERL@ ./mysql-test-run.pl --comment=debug  --force --timer \
-	        --skip-ndbcluster --skip-rpl --report-features
-
 test-bt-debug-fast:
 
 # Keep these for a while
@@ -218,35 +116,8 @@
 test-force-pl-mem:  test-force-mem
 test-force-full-pl: test-force-full
 
-test-ext-funcs:
-	cd mysql-test ; \
-	    @PERL@ ./mysql-test-run.pl --force --reorder --suite=funcs_1 ; \
-	    @PERL@ ./mysql-test-run.pl --force --suite=funcs_2
-
-test-ext-rpl:
-	cd mysql-test ; \
-	    @PERL@ ./mysql-test-run.pl --force --suite=rpl
-
-test-ext-partitions:
-	cd mysql-test ; \
-	    @PERL@ ./mysql-test-run.pl --force --suite=parts
-
-test-ext-jp:
-	cd mysql-test ; \
-	    @PERL@ ./mysql-test-run.pl --force --suite=jp
-
-test-ext-stress:
-	cd mysql-test ; \
-	    @PERL@ ./mysql-test-run.pl --force --big-test --suite=stress
-
 test-ext:	test-ext-funcs test-ext-rpl test-ext-partitions test-ext-jp test-ext-stress
 
-test-fast:
-	cd mysql-test ; \
-	    @PERL@ ./mysql-test-run.pl $(subset) --force --skip-ndb --skip-innodb --skip-im --skip-rpl ; \
-	    @PERL@ ./mysql-test-run.pl $(subset) --force --suite=funcs_1 --do-test=myisam ; \
-	    @PERL@ ./mysql-test-run.pl $(subset) --force --suite=stress --do-test=ddl_myisam 
-
 test-fast-view:
 	$(MAKE) subset=--view-protocol test-fast
 
diff -ur mysql-6.0.9-alpha/Makefile.in mysql-6.0.9-alpha_new/Makefile.in
--- mysql-6.0.9-alpha/Makefile.in	2009-01-02 10:59:49.000000000 -0500
+++ mysql-6.0.9-alpha_new/Makefile.in	2009-01-30 22:36:36.000000000 -0500
@@ -143,18 +143,18 @@
 DIFF = @DIFF@
 DOT_FRM_VERSION = @DOT_FRM_VERSION@
 DOXYGEN = @DOXYGEN@
+DSYMUTIL = @DSYMUTIL@
 DTRACE = @DTRACE@
 DTRACEFLAGS = @DTRACEFLAGS@
-ECHO = @ECHO@
+DUMPBIN = @DUMPBIN@
 ECHO_C = @ECHO_C@
 ECHO_N = @ECHO_N@
 ECHO_T = @ECHO_T@
 EGREP = @EGREP@
 EXEEXT = @EXEEXT@
-F77 = @F77@
 FALCON_CXXFLAGS = @FALCON_CXXFLAGS@
 FALCON_SPARC_ASSEMBLY = @FALCON_SPARC_ASSEMBLY@
-FFLAGS = @FFLAGS@
+FGREP = @FGREP@
 FIND_PROC = @FIND_PROC@
 GETCONF = @GETCONF@
 GREP = @GREP@
@@ -177,6 +177,7 @@
 LIBS = @LIBS@
 LIBTOOL = @LIBTOOL@
 LIB_EXTRA_CCFLAGS = @LIB_EXTRA_CCFLAGS@
+LIPO = @LIPO@
 LM_CFLAGS = @LM_CFLAGS@
 LN = @LN@
 LN_CP_F = @LN_CP_F@
@@ -221,9 +222,13 @@
 NDB_VERSION_MINOR = @NDB_VERSION_MINOR@
 NDB_VERSION_STATUS = @NDB_VERSION_STATUS@
 NM = @NM@
+NMEDIT = @NMEDIT@
 NOINST_LDFLAGS = @NOINST_LDFLAGS@
 NON_THREADED_LIBS = @NON_THREADED_LIBS@
+OBJDUMP = @OBJDUMP@
 OBJEXT = @OBJEXT@
+OTOOL = @OTOOL@
+OTOOL64 = @OTOOL64@
 PACKAGE = @PACKAGE@
 PACKAGE_BUGREPORT = @PACKAGE_BUGREPORT@
 PACKAGE_NAME = @PACKAGE_NAME@
@@ -270,7 +275,7 @@
 abs_top_srcdir = @abs_top_srcdir@
 ac_ct_CC = @ac_ct_CC@
 ac_ct_CXX = @ac_ct_CXX@
-ac_ct_F77 = @ac_ct_F77@
+ac_ct_DUMPBIN = @ac_ct_DUMPBIN@
 am__include = @am__include@
 am__leading_dot = @am__leading_dot@
 am__quote = @am__quote@
@@ -313,6 +318,7 @@
 libmysqld_dirs = @libmysqld_dirs@
 localedir = @localedir@
 localstatedir = @localstatedir@
+lt_ECHO = @lt_ECHO@
 man1_files = @man1_files@
 man8_files = @man8_files@
 man_dirs = @man_dirs@
@@ -347,16 +353,12 @@
 plugin_blackhole_static_target = @plugin_blackhole_static_target@
 plugin_csv_shared_target = @plugin_csv_shared_target@
 plugin_csv_static_target = @plugin_csv_static_target@
-plugin_daemon_example_shared_target = @plugin_daemon_example_shared_target@
-plugin_daemon_example_static_target = @plugin_daemon_example_static_target@
 plugin_example_shared_target = @plugin_example_shared_target@
 plugin_example_static_target = @plugin_example_static_target@
 plugin_falcon_shared_target = @plugin_falcon_shared_target@
 plugin_falcon_static_target = @plugin_falcon_static_target@
 plugin_federated_shared_target = @plugin_federated_shared_target@
 plugin_federated_static_target = @plugin_federated_static_target@
-plugin_ftexample_shared_target = @plugin_ftexample_shared_target@
-plugin_ftexample_static_target = @plugin_ftexample_static_target@
 plugin_heap_shared_target = @plugin_heap_shared_target@
 plugin_heap_static_target = @plugin_heap_static_target@
 plugin_innobase_shared_target = @plugin_innobase_shared_target@
@@ -421,7 +423,7 @@
 			@sql_union_dirs@ unittest storage plugin \
 			@sql_server@ @man_dirs@ tests \
 			netware @libmysqld_dirs@ \
-			mysql-test support-files sql-bench \
+			support-files \
 			win
 
 DIST_SUBDIRS = . include Docs zlib \
@@ -430,7 +432,7 @@
 			strings mysys dbug mysys/tests extra regex libmysql libmysql_r client unittest storage plugin \
 			vio sql man tests \
 			netware libmysqld \
-			mysql-test support-files sql-bench \
+			support-files \
 			win \
 			BUILD
 
@@ -928,41 +930,6 @@
 test-unit-big:
 	cd unittest; MYTAP_CONFIG=big $(MAKE) test
 
-test-ps:
-	cd mysql-test ; \
-	    @PERL@ ./mysql-test-run.pl $(force) --ps-protocol --mysqld=--binlog-format=mixed
-
-test-nr:
-	cd mysql-test ; \
-	    @PERL@ ./mysql-test-run.pl $(force) --mysqld=--binlog-format=row
-
-test-pr:
-	cd mysql-test ; \
-	    @PERL@ ./mysql-test-run.pl $(force) $(mem) --ps-protocol --mysqld=--binlog-format=row #@libevent_test_option@
-
-test-ns:
-	cd mysql-test ; \
-	    @PERL@ ./mysql-test-run.pl $(force) $(mem) --mysqld=--binlog-format=mixed
-
-test-binlog-statement:
-	cd mysql-test ; \
-	    @PERL@ ./mysql-test-run.pl $(force) --mysqld=--binlog-format=statement
-
-# This code is duplicated in "test-bt", see the Changeset comment of 2007-Dec-07
-test-embedded:
-	if [ -e bin/mysqltest_embedded -o -e libmysqld/examples/mysqltest_embedded ] ; then \
-	  cd mysql-test ; MTR_BUILD_THREAD=auto \
-	      @PERL@ ./mysql-test-run.pl --comment=embedded --force --timer \
-	          --embedded-server --skip-rpl --skip-ndbcluster ; \
-	else \
-	  echo "no program found for 'embedded' tests - skipped testing" ; \
-	fi
-
-test-reprepare:
-	cd mysql-test ; \
-	    @PERL@ ./mysql-test-run.pl $(force) $(mem) --ps-protocol \
-		--mysqld=--debug=+d,reprepare_each_statement
-
 test:	test-ns test-pr
 
 test-full:	test test-nr test-ps
@@ -977,73 +944,6 @@
 test-force-mem:
 	$(MAKE) force=--force mem=--mem test
 
-test-bt:
-	-cd mysql-test ; MTR_BUILD_THREAD=auto \
-	    @PERL@ ./mysql-test-run.pl --comment=normal --force --timer \
-	        --skip-ndbcluster --report-features
-	-cd mysql-test ; MTR_BUILD_THREAD=auto \
-	    @PERL@ ./mysql-test-run.pl  --comment=ps    --force --timer \
-	        --skip-ndbcluster --ps-protocol
-	-if [ -e bin/ndbd -o -e storage/ndb/src/kernel/ndbd ] ; then \
-	  cd mysql-test ; \
-	    MTR_BUILD_THREAD=auto \
-	      @PERL@ ./mysql-test-run.pl --comment=ndb+rpl_ndb+ps --force --timer \
-	          --ps-protocol --mysqld=--binlog-format=row --suite=ndb,rpl_ndb ; \
-	    MTR_BUILD_THREAD=auto \
-	      @PERL@ ./mysql-test-run.pl --comment=ndb --force --timer \
-	          --with-ndbcluster-only ; \
-	else \
-	  echo "no program found for 'ndbcluster' tests - skipped testing" ; \
-	fi
-	-cd mysql-test ; MTR_BUILD_THREAD=auto \
-	    @PERL@ ./mysql-test-run.pl --force --comment=falcon --suite=falcon
-	-cd mysql-test ; MTR_BUILD_THREAD=auto \
-	    @PERL@ ./mysql-test-run.pl --force --comment=funcs1+ps --ps-protocol --reorder --suite=funcs_1
-	-cd mysql-test ; MTR_BUILD_THREAD=auto \
-	    @PERL@ ./mysql-test-run.pl --force --comment=funcs2 --suite=funcs_2
-	-cd mysql-test ; MTR_BUILD_THREAD=auto \
-	    @PERL@ ./mysql-test-run.pl --force --comment=partitions --suite=parts
-	-cd mysql-test ; MTR_BUILD_THREAD=auto \
-	    @PERL@ ./mysql-test-run.pl --force --comment=stress --suite=stress
-	-cd mysql-test ; MTR_BUILD_THREAD=auto \
-	    @PERL@ ./mysql-test-run.pl --force --comment=jp --suite=jp
-	-if [ -d mysql-test/suite/nist ] ; then \
-	  cd mysql-test ; MTR_BUILD_THREAD=auto \
-	      @PERL@ ./mysql-test-run.pl --comment=nist --force --suite=nist ; \
-	fi
-	-if [ -d mysql-test/suite/nist ] ; then \
-	  cd mysql-test ; MTR_BUILD_THREAD=auto \
-	      @PERL@ ./mysql-test-run.pl --comment=NIST+normal --force --suite=nist ; \
-	      @PERL@ ./mysql-test-run.pl --comment=NIST+ps --force --suite=nist --ps-protocol ; \
-	fi
-	-if [ -e bin/mysqltest_embedded -o -e libmysqld/examples/mysqltest_embedded ] ; then \
-	  cd mysql-test ; MTR_BUILD_THREAD=auto \
-	      @PERL@ ./mysql-test-run.pl --comment=embedded --force --timer \
-	          --embedded-server --skip-rpl --skip-ndbcluster ; \
-	else \
-	  echo "no program found for 'embedded' tests - skipped testing" ; \
-	fi
-
-test-bt-fast:
-	-cd mysql-test ; MTR_BUILD_THREAD=auto \
-	    @PERL@ ./mysql-test-run.pl  --comment=ps    --force --timer \
-	        --skip-ndbcluster --ps-protocol --report-features
-	-if [ -e bin/ndbd -o -e storage/ndb/src/kernel/ndbd ] ; then \
-	  cd mysql-test ; \
-	    MTR_BUILD_THREAD=auto \
-	      @PERL@ ./mysql-test-run.pl --comment=ndb --force --timer \
-	          --with-ndbcluster-only ; \
-	else \
-	  echo "no program found for 'ndbcluster' tests - skipped testing" ; \
-	fi
-	-cd mysql-test ; MTR_BUILD_THREAD=auto \
-	    @PERL@ ./mysql-test-run.pl --force --comment=stress --suite=stress
-
-test-bt-debug:
-	-cd mysql-test ; MTR_BUILD_THREAD=auto \
-	    @PERL@ ./mysql-test-run.pl --comment=debug  --force --timer \
-	        --skip-ndbcluster --skip-rpl --report-features
-
 test-bt-debug-fast:
 
 # Keep these for a while
@@ -1053,35 +953,8 @@
 test-force-pl-mem:  test-force-mem
 test-force-full-pl: test-force-full
 
-test-ext-funcs:
-	cd mysql-test ; \
-	    @PERL@ ./mysql-test-run.pl --force --reorder --suite=funcs_1 ; \
-	    @PERL@ ./mysql-test-run.pl --force --suite=funcs_2
-
-test-ext-rpl:
-	cd mysql-test ; \
-	    @PERL@ ./mysql-test-run.pl --force --suite=rpl
-
-test-ext-partitions:
-	cd mysql-test ; \
-	    @PERL@ ./mysql-test-run.pl --force --suite=parts
-
-test-ext-jp:
-	cd mysql-test ; \
-	    @PERL@ ./mysql-test-run.pl --force --suite=jp
-
-test-ext-stress:
-	cd mysql-test ; \
-	    @PERL@ ./mysql-test-run.pl --force --big-test --suite=stress
-
 test-ext:	test-ext-funcs test-ext-rpl test-ext-partitions test-ext-jp test-ext-stress
 
-test-fast:
-	cd mysql-test ; \
-	    @PERL@ ./mysql-test-run.pl $(subset) --force --skip-ndb --skip-innodb --skip-im --skip-rpl ; \
-	    @PERL@ ./mysql-test-run.pl $(subset) --force --suite=funcs_1 --do-test=myisam ; \
-	    @PERL@ ./mysql-test-run.pl $(subset) --force --suite=stress --do-test=ddl_myisam 
-
 test-fast-view:
 	$(MAKE) subset=--view-protocol test-fast
 
diff -ur mysql-6.0.9-alpha/configure mysql-6.0.9-alpha_new/configure
--- mysql-6.0.9-alpha/configure	2009-01-02 10:59:54.000000000 -0500
+++ mysql-6.0.9-alpha_new/configure	2009-01-30 22:41:18.000000000 -0500
@@ -56041,7 +56041,7 @@
 
 fi
 
-ac_config_files="$ac_config_files Makefile extra/Makefile mysys/Makefile mysys/tests/Makefile unittest/Makefile unittest/mytap/Makefile unittest/mytap/t/Makefile unittest/mysys/Makefile unittest/examples/Makefile strings/Makefile regex/Makefile storage/Makefile man/Makefile BUILD/Makefile vio/Makefile libmysql/Makefile libmysql_r/Makefile client/Makefile sql/Makefile sql/share/Makefile sql/backup/Makefile sql/sql_builtin.cc sql-common/Makefile dbug/Makefile scripts/Makefile include/Makefile tests/Makefile Docs/Makefile support-files/Makefile support-files/MacOSX/Makefile support-files/RHEL4-SElinux/Makefile cmd-line-utils/Makefile cmd-line-utils/libedit/Makefile libmysqld/Makefile libmysqld/examples/Makefile mysql-test/Makefile mysql-test/ndb/Makefile netware/Makefile sql-bench/Makefile include/mysql_version.h plugin/Makefile win/Makefile"
+ac_config_files="$ac_config_files Makefile extra/Makefile mysys/Makefile mysys/tests/Makefile unittest/Makefile unittest/mytap/Makefile unittest/mytap/t/Makefile unittest/mysys/Makefile unittest/examples/Makefile strings/Makefile regex/Makefile storage/Makefile man/Makefile BUILD/Makefile vio/Makefile libmysql/Makefile libmysql_r/Makefile client/Makefile sql/Makefile sql/share/Makefile sql/backup/Makefile sql/sql_builtin.cc sql-common/Makefile dbug/Makefile scripts/Makefile include/Makefile tests/Makefile Docs/Makefile support-files/Makefile support-files/MacOSX/Makefile support-files/RHEL4-SElinux/Makefile cmd-line-utils/Makefile cmd-line-utils/libedit/Makefile libmysqld/Makefile libmysqld/examples/Makefile netware/Makefile include/mysql_version.h plugin/Makefile win/Makefile"
 
 
 ac_config_commands="$ac_config_commands default"
@@ -56875,10 +56875,7 @@
     "cmd-line-utils/libedit/Makefile") CONFIG_FILES="$CONFIG_FILES cmd-line-utils/libedit/Makefile" ;;
     "libmysqld/Makefile") CONFIG_FILES="$CONFIG_FILES libmysqld/Makefile" ;;
     "libmysqld/examples/Makefile") CONFIG_FILES="$CONFIG_FILES libmysqld/examples/Makefile" ;;
-    "mysql-test/Makefile") CONFIG_FILES="$CONFIG_FILES mysql-test/Makefile" ;;
-    "mysql-test/ndb/Makefile") CONFIG_FILES="$CONFIG_FILES mysql-test/ndb/Makefile" ;;
     "netware/Makefile") CONFIG_FILES="$CONFIG_FILES netware/Makefile" ;;
-    "sql-bench/Makefile") CONFIG_FILES="$CONFIG_FILES sql-bench/Makefile" ;;
     "include/mysql_version.h") CONFIG_FILES="$CONFIG_FILES include/mysql_version.h" ;;
     "plugin/Makefile") CONFIG_FILES="$CONFIG_FILES plugin/Makefile" ;;
     "win/Makefile") CONFIG_FILES="$CONFIG_FILES win/Makefile" ;;
diff -ur mysql-6.0.9-alpha/configure.in mysql-6.0.9-alpha_new/configure.in
--- mysql-6.0.9-alpha/configure.in	2009-01-02 10:45:40.000000000 -0500
+++ mysql-6.0.9-alpha_new/configure.in	2009-01-30 22:41:18.000000000 -0500
@@ -2836,8 +2836,7 @@
  support-files/MacOSX/Makefile support-files/RHEL4-SElinux/Makefile dnl
  cmd-line-utils/Makefile cmd-line-utils/libedit/Makefile dnl
  libmysqld/Makefile libmysqld/examples/Makefile dnl
- mysql-test/Makefile dnl
- mysql-test/ndb/Makefile netware/Makefile sql-bench/Makefile dnl
+ netware/Makefile dnl
  include/mysql_version.h plugin/Makefile win/Makefile)
 
 AC_CONFIG_COMMANDS([default], , test -z "$CONFIG_HEADERS" || echo timestamp > stamp-h)
