diff -Naur dmraid.org/1.0.0.rc11/lib/Makefile.in dmraid/1.0.0.rc11/lib/Makefile.in
--- dmraid.org/1.0.0.rc11/lib/Makefile.in	2006-02-17 16:19:25.000000000 +0000
+++ dmraid/1.0.0.rc11/lib/Makefile.in	2006-04-10 21:47:38.000000000 +0100
@@ -35,7 +35,6 @@
 	format/ataraid/pdc.c \
 	format/ataraid/sil.c \
 	format/ataraid/via.c \
-	format/ataraid/asr.c \
 	format/partition/dos.c
 
 OBJECTS=$(SOURCES:%.c=%.o)
diff -Naur dmraid.org/1.0.0.rc11/lib/Makefile.in.orig dmraid/1.0.0.rc11/lib/Makefile.in.orig
--- dmraid.org/1.0.0.rc11/lib/Makefile.in.orig	2006-02-17 16:05:42.000000000 +0000
+++ dmraid/1.0.0.rc11/lib/Makefile.in.orig	2006-04-10 21:47:50.000000000 +0100
@@ -35,7 +35,6 @@
 	format/ataraid/pdc.c \
 	format/ataraid/sil.c \
 	format/ataraid/via.c \
-	format/ataraid/asr.c \
 	format/partition/dos.c
 
 OBJECTS=$(SOURCES:%.c=%.o)

